### PR TITLE
Add a Task class specialised for Django

### PR DESCRIPTION
**Pull request by browniebroke**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryq.dev/en/main/contributing.html).

## Description

Add a task specialised to use with Django. Offers an API to call `.delay()` on tasks at the end of the database transaction. 

- [x] Add task to the codebase
- [x] Add docs
- [x] Add tests 

## Rationale

A common pitfall using Celery with Django is to call `.delay()` in a middle of a database transaction and have the task start executing before the transaction is committed. I think it's partly due to the fact that doing the call properly is a bit cumbersome. 

This PR adds a specialised task that aim at offering a nicer API for it. I've added the initial code and some docs to explain how it would be used, but before going further, I'd like to know whether this would be a welcome addition or if it's considered too out of scope.

### CODE REVIEW

File: celery/contrib/django/task.py
Comment (browniebroke) [inline celery/contrib/django/task.py:21]: I went with some separate methods, to be explicit and in case folks need access to the original ones but maybe it's ok to always do that

File: docs/django/first-steps-with-django.rst
Comment (browniebroke) [inline docs/django/first-steps-with-django.rst:198]: One thing to explore is whether this could be automatic by fiddling in `celery/fixups/django.py`

Comment (auvipy) [reply docs/django/first-steps-with-django.rst:198]: yeah exploring the option to make part of it automatic and configurable/extensible might be a good idea

File: celery/fixups/django.py
Comment (browniebroke) [inline celery/fixups/django.py:80]: One problem at the moment (I think) is that it would override a custom class set by the user.

File: celery/contrib/django/task.py
Comment (auvipy) [reply celery/contrib/django/task.py:21]: I want to be sure that these will work from django 2.2 LTS to 4.2 LTS for a while?

File: celery/fixups/django.py
Comment (auvipy) [reply celery/fixups/django.py:80]: do you have any alternative in mind?

**Review by auvipy — CHANGES_REQUESTED** [review]

File: celery/fixups/django.py
Comment (browniebroke) [reply celery/fixups/django.py:80]: Not yet. I'm thinking that, prior to changing it, I could check the current value and don't change it if it's different than the default task. I don't think there is a use case for people overriding this specialised class back to the default one currently.

Comment (browniebroke) [reply celery/fixups/django.py]: I think I've fixed it now.

File: celery/contrib/django/task.py
Comment (auvipy) [reply celery/contrib/django/task.py:21]: that makes sense thanks.

File: celery/contrib/django/task.py
Comment (auvipy) [inline celery/contrib/django/task.py:8]: should we use the name as Task or DjangoTask or anything better?

Comment (browniebroke) [reply celery/contrib/django/task.py:8]: I was asklng myself the same question. 

One the one hand, this is namespaced into a `contrib.django` package, so one could make the argument that naming the class `DjangoTask` might feel redundant. 

But on the other hand, this namespace might be glossed over or even not visible in logs/stacktraces. Might also help tooling to have a distinct class name. I'm also already doing some `import Task as DjangoTask` in the tests, so I see that as a signal that a different name might make sense.

Given all that, I'm leaning towards the name `DjangoTask`.

File: docs/django/first-steps-with-django.rst
Comment (browniebroke) [inline docs/django/first-steps-with-django.rst:156]: Should I mention a `.. versionadded::` in this part?

Comment (auvipy) [reply docs/django/first-steps-with-django.rst:156]: I think that would help

Comment (browniebroke) [reply docs/django/first-steps-with-django.rst:156]: As it's more of a new feature than a bugfix, I'm assuming that it'll be part of 5.4 rather than the next 5.3.x

File: celery/contrib/django/task.py
Comment (browniebroke) [inline celery/contrib/django/task.py:6]: Added this try/catch to let us generate the API reference page. 

I'm not a big fan of this as it might hide some misconfiguration in the app, the error will happen when the method is called rather than at start-up. 

Is there any pre-existing way of dealing with 3rd party libs that are not present when generating the documentation?

**Comment by thedrow** [comment]
"A common pitfall using Celery with Django is to call .delay() in a middle of a database transaction and have the task start executing before the transaction is committed. I think it's partly due to the fact that doing the call properly is a bit cumbersome."
I'm not sure I'm following here.
If you delay a task in the middle of a database transaction that may or may not be desired.
A delayed task does not participate in the same transaction as the producer so why should we commit before delaying?

**Comment by browniebroke** [comment]
> "A common pitfall using Celery with Django is to call .delay() in a middle of a database transaction and have the task start executing before the transaction is committed. I think it's partly due to the fact that doing the call properly is a bit cumbersome."
> I'm not sure I'm following here.
> If you delay a task in the middle of a database transaction that may or may not be desired.
> A delayed task does not participate in the same transaction as the producer so why should we commit before delaying?

I might not have phrased it in the best way possible (especially in the PR description). I would like it to be well explained in the docs though, and I'm trying to show it with a simple example. Let me know if that helps... 

You're right: it may or may not be desired, but in my experience, it was not desired in the vast majority of the cases.

**Review by federicobond — APPROVED** [review]
This is a very useful change. I have written helpers to achieve the same behavior in my projects.

File: celery/contrib/django/task.py
Comment (auvipy) [inline celery/contrib/django/task.py:8]:
```suggestion
from celery.app.task import Task
```

File: celery/contrib/django/task.py
Comment (auvipy) [inline celery/contrib/django/task.py:11]:
```suggestion
class DjangoTask(Task):
```

**Review by auvipy — CHANGES_REQUESTED** [review]

**Comment by Nusnus** [comment]
@browniebroke hey there - thanks for this contribution, I'm sure it will be very useful!

Do you think it's possible to also refine the django example or add another one for the new feature?

I'm positive it would increase adoption and possibly awareness of the new feature.

**Comment by Nusnus** [comment]
@browniebroke hey there - thanks for this contribution, I'm sure it will be very useful!

Do you think it's possible to also refine the django example or add another one for the new feature?

I'm positive it would increase adoption and possibly awareness of the new feature.

**Comment by browniebroke** [comment]
> Do you think it's possible to also refine the django example or add another one for the new feature?

Absolutely! I actually saw the example at some point during implementation and made a mental note to look at it later, but I (obviously) forgot about it 😄

Thanks for bringing it up.

**Comment by Nusnus** [comment]
> > Do you think it's possible to also refine the django example or add another one for the new feature?
> 
> Absolutely! I actually saw the example at some point during implementation and made a mental note to look at it later, but I (obviously) forgot about it 😄
> 
> Thanks for bringing it up.

Awesome!
Thank you 👑

**Comment by browniebroke** [comment]
> Do you think it's possible to also refine the django example or add another one for the new feature?

I looked at the example and it was actually not making much use of the ``delay()`` API, except in the readme. I've updated it, but I'm thinking that maybe we might want to leave the existing example and write a more relevant example, with a DB transaction. I see that there is a model, so maybe I could make use of that?

File: celery/contrib/django/task.py
Comment (browniebroke) [reply celery/contrib/django/task.py:6]: Amazing, thanks for the pointer!

**Comment by auvipy** [comment]
> > Do you think it's possible to also refine the django example or add another one for the new feature?
> 
> I looked at the example project and it was actually not making much use of the `delay()` API, except in the readme. I've updated it, but I'm thinking that maybe we might want to leave the existing example and write a more relevant example, with a DB transaction. I see that there is a model, so maybe I could make use of that?

yes you can go for it. use that model and use a more nice example....

**Comment by cclauss** [comment]
Nice!!  Perhaps this needs revisiting now that Django v5.0.1 is shipping.

**Review by auvipy — COMMENTED** [review]
do you plan to add anything more to this PR?

File: docs/django/first-steps-with-django.rst
Comment (auvipy) [reply docs/django/first-steps-with-django.rst:156]: yes 5.4 it should be

**Comment by browniebroke** [comment]
No, I don't think so.

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
I will wait for other team mates opinion before merge

**Comment by Nusnus** [comment]
> I will wait for other team mates opinion before merge

I'll check with @thedrow if he's available for another review. Will do myself as well.

**Comment by auvipy** [comment]
> No, I don't think so.

this failed only after merged. it was OK during last run in the PR

tox: 3.9-smoke
  3.9-smoke: 39864 W commands[0]> pytest -xsv t/smoke --dist=loadscope --reruns 10 --reruns-delay 60 --rerun-except AssertionError -n auto -k test_tasks.py [tox/tox_env/api.py:427]
  ============================= test session starts ==============================
  platform linux -- Python 3.9.18, pytest-7.4.4, pluggy-1.4.0 -- /home/runner/work/celery/celery/.tox/3.9-smoke/bin/python
  cachedir: .tox/3.9-smoke/.pytest_cache
  rootdir: /home/runner/work/celery/celery
  configfile: pyproject.toml
  plugins: order-1.2.0, cov-4.1.0, timeout-2.2.0, docker-tools-3.1.3, subtests-0.11.0, click-1.1.0, xdist-3.5.0, github-actions-annotate-failures-0.2.0, celery-1.0.0b1, rerunfailures-13.0
  created: 2/2 workers
  2 workers [18 items]
  
  scheduling tests via LoadScopeScheduling
  
  t/smoke/tests/test_tasks.py::test_replace::test_sanity[celery_setup_worker-celery_redis_broker-celery_redis_backend] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.SIGKILL-WorkerLostError] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.SIGKILL-WorkerLostError] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.SYSTEM_EXIT-WorkerLostError] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.SYSTEM_EXIT-WorkerLostError] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.DELAY_TIMEOUT-TimeLimitExceeded] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.DELAY_TIMEOUT-TimeLimitExceeded] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.EXHAUST_MEMORY-WorkerLostError] 
  [gw1] PASSED t/smoke/tests/test_tasks.py::test_replace::test_sanity[celery_setup_worker-celery_redis_broker-celery_redis_backend] 
  t/smoke/tests/test_tasks.py::test_replace::test_sanity[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_redis_broker-celery_redis_backend-Method.EXHAUST_MEMORY-WorkerLostError] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.SIGKILL-WorkerLostError] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.SIGKILL-WorkerLostError] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.SYSTEM_EXIT-WorkerLostError] 
  [gw1] PASSED t/smoke/tests/test_tasks.py::test_replace::test_sanity[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.SYSTEM_EXIT-WorkerLostError] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.DELAY_TIMEOUT-TimeLimitExceeded] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.DELAY_TIMEOUT-TimeLimitExceeded] 
  t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.EXHAUST_MEMORY-WorkerLostError] 
  [gw0] PASSED t/smoke/tests/test_tasks.py::test_task_termination::test_child_process_respawn[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend-Method.EXHAUST_MEMORY-WorkerLostError] 
  Error: The action has timed out.

**Comment by browniebroke** [comment]
I saw the error and was wondering if it was some flakyness? Seems like the smoke tests worked on other python versions... 🤔 

Do we know where this comes from? `Error: The action has timed out.`

**Comment by Nusnus** [comment]
> I saw the error and was wondering if it was some flakyness? Seems like the smoke tests worked on other python versions... 🤔
> 
> Do we know where this comes from? `Error: The action has timed out.`

Checkout the latest run. It was indeed a hiccup. I already restarted the CI yesterday :)
It passes 100%.

**Comment by Nusnus** [comment]
> > I saw the error and was wondering if it was some flakyness? Seems like the smoke tests worked on other python versions... 🤔
> > Do we know where this comes from? `Error: The action has timed out.`
> 
> Checkout the latest run. It was indeed a hiccup. I already restarted the CI yesterday :) It passes 100%.

hiccup a.k.a Redis container hiccups causes Celery worker to hang -> CI reaches timeout.

**Comment by faradayyg** [comment]
Thanks for this improvement!

**Comment by maciej-gol** [comment]
This broke cases when `task_cls` is defined as an attribute on a custom Celery app, for example: https://github.com/maciej-gol/tenant-schemas-celery/blob/master/tenant_schemas_celery/app.py#L79

```python
class CeleryApp(Celery):
    registry_cls = 'tenant_schemas_celery.registry:TenantTaskRegistry'
    task_cls = 'tenant_schemas_celery.task:TenantTask'
```

The check only looks at the value passed in the c-tor, instead of the attribute being overriden, too.

**Comment by browniebroke** [comment]
Oh I didn't think about this use case. I think it should be supported out of the box. Do you think you could write a patch for Celery? If not, could at least open an issue for it, and @ me into it, please? This description is clear enough to me, I think.

**Comment by maciej-gol** [comment]
@browniebroke I will look into preparing an MR, will let you know if I fail :)

**Comment by browniebroke** [comment]
@maciej-gol started a fix here: https://github.com/celery/celery/pull/9038

**Pull request merged**
